File: packages/backstage-ts/src/logger.ts
/**
 * Backstage SDK - Logger
 *
 * Winston-based logging with optional disable and custom handlers.
 */

import winston from 'winston';

export enum LogLevel {
  DEBUG = 'debug',
  INFO = 'info',
  WARN = 'warn',
  ERROR = 'error',
}

export interface LogEntry {
  level: LogLevel;
  message: string;
  timestamp: string;
  payload?: Record<string, unknown>;
  error?: string;
  stack?: string;
}

export type LogHandler = (entry: LogEntry) => void | Promise<void>;

export interface LoggerConfig {
  /** Minimum log level to output */
  level?: LogLevel;
  /** File path to write logs to (JSON format) */
  file?: string;
  /** Whether this is a child logger */
  isChild?: boolean;
  /** Whether this is a scheduler logger (changes prefix) */
  isScheduler?: boolean;
  /** Custom log handler function */
  handler?: LogHandler;
  /** Disable all logging */
  silent?: boolean;
}

/**
 * Logger wrapper around Winston.
 * Supports console output, file output, and custom handlers.
 */
export class Logger {
  private winston: winston.Logger;
  private handler?: LogHandler;
  private silent: boolean;
  private isChild: boolean;
  private prefix: string;

  /**
   * Create a new Logger instance.
   *
   * @param config - Logger configuration
   */
  constructor(config: LoggerConfig = {}) {
    this.handler = config.handler;
    this.silent = config.silent ?? false;
    this.isChild = config.isChild ?? false;
    this.prefix = config.isScheduler ? 'Scheduler' : 'Worker';

    const transports: winston.transport[] = [];

    if (!this.silent) {
      transports.push(
        new winston.transports.Console({
          format: winston.format.combine(
            winston.format.timestamp(),
            winston.format.printf(({ timestamp, level, message }) => {
              return `${timestamp} ${level.toUpperCase().padEnd(5)} [${
                this.prefix
              }]: ${message}`;
            }),
          ),
        }),
      );
    }

    if (config.file) {
      transports.push(
        new winston.transports.File({
          filename: config.file,
          format: winston.format.combine(
            winston.format.timestamp(),
            winston.format.json(),
          ),
        }),
      );
    }

    this.winston = winston.createLogger({
      level: config.level ?? LogLevel.INFO,
      silent: this.silent && !config.file && !this.handler,
      transports,
    });
  }

  /**
   * Set a custom log handler.
   * Useful for sending logs to external services.
   */
  setHandler(handler: LogHandler): void {
    this.handler = handler;
  }

  /**
   * Silence the logger at runtime.
   */
  setSilent(silent: boolean): void {
    this.silent = silent;
    this.winston.silent = silent && !this.handler;
  }

  private callHandler(entry: LogEntry): void {
    if (this.handler) {
      try {
        this.handler(entry);
      } catch {
        // Don't let handler errors break logging
      }
    }
  }

  /**
   * Log a debug message.
   */
  debug(message: string, payload?: Record<string, unknown>): void {
    const entry: LogEntry = {
      level: LogLevel.DEBUG,
      message,
      timestamp: new Date().toISOString(),
      payload,
    };
    this.callHandler(entry);
    this.winston.debug(message, { payload });
  }

  /**
   * Log an info message.
   */
  info(message: string, payload?: Record<string, unknown>): void {
    const entry: LogEntry = {
      level: LogLevel.INFO,
      message,
      timestamp: new Date().toISOString(),
      payload,
    };
    this.callHandler(entry);
    this.winston.info(message, { payload });
  }

  /**
   * Log a warning message.
   */
  warn(message: string, payload?: Record<string, unknown>): void {
    const entry: LogEntry = {
      level: LogLevel.WARN,
      message,
      timestamp: new Date().toISOString(),
      payload,
    };
    this.callHandler(entry);
    this.winston.warn(message, { payload });
  }

  /**
   * Log an error message.
   * Accepts a string or an Error object.
   */
  error(message: string | Error, payload?: Record<string, unknown>): void {
    if (message instanceof Error) {
      const entry: LogEntry = {
        level: LogLevel.ERROR,
        message: message.message,
        timestamp: new Date().toISOString(),
        error: message.name,
        stack: message.stack,
        payload,
      };
      this.callHandler(entry);
      this.winston.error(message.message, {
        error: message.name,
        stack: message.stack,
        payload,
      });
    } else {
      const entry: LogEntry = {
        level: LogLevel.ERROR,
        message,
        timestamp: new Date().toISOString(),
        payload,
      };
      this.callHandler(entry);
      this.winston.error(message, { payload });
    }
  }

  async flush(): Promise<void> {
    // Winston handles flushing internally
  }

  async close(): Promise<void> {
    this.winston.close();
  }
}

/**
 * Helper to create a logger instance.
 */
export function createLogger(config?: LoggerConfig): Logger {
  return new Logger(config);
}

File: packages/backstage-go/logger.go
// Package backstage logger.
// Provides structured logging using log/slog.
package backstage

import (
	"context"
	"io"
	"log/slog"
	"os"
)

// LogHandler is the interface for custom log handlers.
type LogHandler func(level slog.Level, msg string, attrs ...slog.Attr)

type Logger struct {
	slog    *slog.Logger
	handler LogHandler
	silent  bool
}

type LoggerConfig struct {
	Level   slog.Level
	Handler LogHandler
	Silent  bool
	Output  io.Writer
}

func NewLogger(prefix string, config ...LoggerConfig) *Logger {
	cfg := LoggerConfig{Level: slog.LevelInfo}
	if len(config) > 0 {
		cfg = config[0]
	}

	output := cfg.Output
	if output == nil {
		output = os.Stdout
	}

	opts := &slog.HandlerOptions{Level: cfg.Level}
	var slogHandler slog.Handler

	if cfg.Silent && cfg.Handler == nil {
		slogHandler = slog.NewTextHandler(io.Discard, opts)
	} else {
		slogHandler = slog.NewTextHandler(output, opts)
	}

	return &Logger{
		slog:    slog.New(slogHandler).With("component", prefix),
		handler: cfg.Handler,
		silent:  cfg.Silent,
	}
}

func (l *Logger) SetHandler(handler LogHandler) {
	l.handler = handler
}

func (l *Logger) SetSilent(silent bool) {
	l.silent = silent
}

func (l *Logger) Debug(msg string, args ...any) {
	if l.handler != nil {
		l.handler(slog.LevelDebug, msg)
	}
	if !l.silent {
		l.slog.Debug(msg, args...)
	}
}

func (l *Logger) Info(msg string, args ...any) {
	if l.handler != nil {
		l.handler(slog.LevelInfo, msg)
	}
	if !l.silent {
		l.slog.Info(msg, args...)
	}
}

func (l *Logger) Warn(msg string, args ...any) {
	if l.handler != nil {
		l.handler(slog.LevelWarn, msg)
	}
	if !l.silent {
		l.slog.Warn(msg, args...)
	}
}

func (l *Logger) Error(msg string, args ...any) {
	if l.handler != nil {
		l.handler(slog.LevelError, msg)
	}
	if !l.silent {
		l.slog.Error(msg, args...)
	}
}

func (l *Logger) With(args ...any) *Logger {
	return &Logger{
		slog:    l.slog.With(args...),
		handler: l.handler,
		silent:  l.silent,
	}
}

// LogEntry for structured logging compatibility
type LogEntry struct {
	Level   string
	Message string
	Attrs   map[string]any
}

// DefaultLogger for package-level logging
var DefaultLogger = NewLogger("Backstage")

// Context-aware logging
func (l *Logger) DebugContext(ctx context.Context, msg string, args ...any) {
	l.slog.DebugContext(ctx, msg, args...)
}

func (l *Logger) InfoContext(ctx context.Context, msg string, args ...any) {
	l.slog.InfoContext(ctx, msg, args...)
}

func (l *Logger) WarnContext(ctx context.Context, msg string, args ...any) {
	l.slog.WarnContext(ctx, msg, args...)
}

func (l *Logger) ErrorContext(ctx context.Context, msg string, args ...any) {
	l.slog.ErrorContext(ctx, msg, args...)
}
